### PR TITLE
feat(kuma-cp) envoy configs for fault injections

### PR DESCRIPTION
**Pull request by lobkovilya**
### Summary

Final major part for implementing Fault Injections. Add configurers to envoy configuration. 
Based on: https://github.com/Kong/kuma/pull/647

### CODE REVIEW

File: pkg/xds/envoy/listeners/fault_injection_configurer.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+				ConfigType: &envoy_hcm.HttpFilter_TypedConfig{
+					TypedConfig: pbst,
+				},
+			},
+		}, manager.HttpFilters...)
+		return nil
+	})
+}
+
+func createHeaders(tagSets []v1alpha1.SingleValueTagSet) (matchers []*envoy_api_v2_route.HeaderMatcher) {
```
Comment (jakubdyszkiewicz) [inline pkg/xds/envoy/listeners/fault_injection_configurer.go]: I might be reading something wrong, but I think there is a problem here. When I specify such policy
```
sources:
- service: payments
- service: ratings
destination:
- redis
conf:
  delay:
    value: 10ms
    percentage: 100.0
```
I want to slow down the connection between payments->redis and ratings->redis.

Multiple matchers in the sources section are OR not AND.
That means that we have to generate Fault Injection filter for every source matcher object

File: pkg/xds/envoy/listeners/util.go [hunk 2]
```diff
@@ -63,3 +70,63 @@ func UpdateFilterConfig(filterChain *envoy_listener.FilterChain, filterName stri
...
+
+	return &envoy_type.FractionalPercent{
+		Numerator:   uint32(math.Round(million * value)),
+		Denominator: envoy_type.FractionalPercent_MILLION,
+	}
+}
+
+var bandwidthRegex = regexp.MustCompile(`(\d*)\s?([gmk]?bps)`)
+
+func ConvertBandwidth(bandwidth *wrappers.StringValue) (*envoy_filter_fault.FaultRateLimit_FixedLimit_, error) {
```
Comment (jakubdyszkiewicz) [inline pkg/xds/envoy/listeners/util.go]: This function works only for Fault Injection (`envoy_filter_fault.FaultRateLimit_FixedLimit_`) so it's not util in any way. It should be placed in fault configurer.

File: pkg/xds/envoy/routes/reset_tags_configurer.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package routes
```
Comment (jakubdyszkiewicz) [inline pkg/xds/envoy/routes/reset_tags_configurer.go]: If you use the code from other PR, please open a PR against that PR. It makes the code review easier because I don't have to reread the code.

So this PR should not be to `master` but to `feat/header-with-tags`, then when `feat/header-with-tags` is merged to master you can rebase and change the target of this PR to `master`

File: pkg/xds/envoy/listeners/fault_injection_configurer.go (hunk 1 above)
Comment (lobkovilya) [reply pkg/xds/envoy/listeners/fault_injection_configurer.go]: Every fault injection filter has multiple headers matchers, which work as OR. So we just map every source matcher on every header matcher. So I guess case you described will work

Comment (jakubdyszkiewicz) [reply pkg/xds/envoy/listeners/fault_injection_configurer.go]: Sorry about that, you are right! Can we extend the `fault_injection_configurer_test.go` so we have this case covered?

One more thing. Are we supporting `*` in regex?
you can specify such policy
```
sources:
- service: *
```
or
```
sources:
- service: *
  cloud: aws
```

Can you please also add test cases for it?

Comment (lobkovilya) [reply pkg/xds/envoy/listeners/fault_injection_configurer.go]: Yes, we are supporting `*`. There is already test cases for this: https://github.com/Kong/kuma/pull/649/files#diff-9c5af35e9a515658d55c693807d6d892R69

Comment (lobkovilya) [reply pkg/xds/envoy/listeners/fault_injection_configurer.go]: > Can we extend the fault_injection_configurer_test.go so we have this case covered?

Sure, I will add test for it

**Comment by jakubdyszkiewicz** [comment]
I think there is one more edge case with this solution.

Let's assume there is dataplane with 2 inbounds
```
inbounds:
- port: 8080
  tags:
    service: web
- port: 8081
  tags:
    service: web-api
```

and we've got such policy
```
sources:
- service: edge
destinations:
- service: web
conf:
...
```

I think this will be applied on both edge->web and edge->web-api and not only on edge->web-api.

The logic the matching logic should be exactly the same with TrafficPermission as with FaultInjection.
You should run the matching logic for every inbound interface. Keep in mind that TrafficPermission seems to work with addition (apply all that matches), not the best match right now (it was the first policy, I think we had a different concept in mind then and it's not unified).

File: pkg/core/policy/dataplane_matcher.go [hunk 4]
```diff
@@ -4,16 +4,19 @@ import (
...
+}
+
+// SelectDataplanePolicy given a something that could be matched (mesh_proto.Dataplane,
+// mesh_proto.Dataplane_Networking_Inbound, etc.) and a list of DataplanePolicy returns the "best matching" DataplanePolicy.
 // Every matching DataplanePolicy gets a rank (score) defined as a maximum number of tags in a matching selector.
 // DataplanePolicy with an empty list of selectors is considered a match with a rank (score) of 0.
 // DataplanePolicy with an empty selector (one that has no tags) is considered a match with a rank (score) of 0.
 // In case if there are multiple DataplanePolicies with the same rank (score), the policy created last is chosen.
-func SelectDataplanePolicy(dataplane *mesh.DataplaneResource, policies []DataplanePolicy) DataplanePolicy {
+func SelectDataplanePolicy(matching MatchingTags, policies []DataplanePolicy) DataplanePolicy {
```
Comment (jakubdyszkiewicz) [inline pkg/core/policy/dataplane_matcher.go:19]: This is clever solution, but can we take a step back?

The solution with reusing DataplanePolicy for matching ConnectionPolicy with adapter is DRY, but I must say this is a little bit hard to follow. This is one of the most important logic in our code, so I want it to be as understandable as possible.

I'd rather have simple understandable interfaces with clear input. I prefer to have a little bit of copying instead of using wrong abstraction (https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction)

Additionally, the code from FaultInjectionMatcher should be in `policy` since it will be soon used for TrafficPermission and also sooner or later on TrafficLogs, so `FaultInjectionMatcher` should be as thin as possible.

I experimented on branch and here is the result https://github.com/Kong/kuma/commit/bc37646d9a9943f0ec9e870267c5142c99d18c64
we don't need te full logic from DataplaneMatcher because with ConnectionPolicies we enforce that there are sources and destinations.

What do you think?

File: pkg/xds/envoy/tags/serialize.go [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+	mesh_proto "github.com/Kong/kuma/api/mesh/v1alpha1"
+)
+
+func Serialize(tags mesh_proto.MultiValueTagSet) string {
+	var pairs []string
+	for _, key := range tags.Keys() {
+		pairs = append(pairs, fmt.Sprintf("&%s=%s&", key, strings.Join(tags.Values(key), ",")))
+	}
+	if len(pairs) == 0 {
```
Comment (jakubdyszkiewicz) [inline pkg/xds/envoy/tags/serialize.go:15]: is this if necessary? 
https://play.golang.org/p/CTr0tpuPYDs

File: pkg/xds/envoy/tags/match_test.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+			expected: true,
+		}),
+		Entry("match without middle tag2", testCase{
+			serviceTags: mesh_proto.MultiValueTagSet{
+				"tag1": {"value1": true, "value2": true},
+				"tag2": {"value2": true, "value3": true},
+				"tag3": {"value3": true, "value4": true},
+			},
+			selector: mesh_proto.SingleValueTagSet{
+				"tag1": "value1",
```
Comment (jakubdyszkiewicz) [inline pkg/xds/envoy/tags/match_test.go:52]: can we change this to `value2` just to see if we match later tag from MultiValuteTagSet?

File: pkg/xds/envoy/listeners/fault_injection_configurer.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	if f.faultInjection == nil {
+		return nil
+	}
+
+	config := &envoy_http_fault.HTTPFault{
+		Delay: convertDelay(f.faultInjection.Conf.GetDelay()),
+		Abort: convertAbort(f.faultInjection.Conf.GetAbort()),
+		Headers: []*envoy_api_v2_route.HeaderMatcher{
+			createHeaders(f.faultInjection.SourceTags()),
+		},
```
Comment (jakubdyszkiewicz) [inline pkg/xds/envoy/listeners/fault_injection_configurer.go:41]: Please remind me. Why can't we generate separate `envoy_api_v2_route.HeaderMatcher` for every selector.Match from FaultInjection? This would simplify the logic of constructing gigantic regex with ORs to construct regex for every selector.Match

File: pkg/xds/envoy/listeners/fault_injection_configurer.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+		}, manager.HttpFilters...)
+		return nil
+	})
+}
+
+func createHeaders(selectors []mesh_proto.SingleValueTagSet) *envoy_api_v2_route.HeaderMatcher {
+	var rss []string
+	for _, selector := range selectors {
+		rs := tags.MatchingRegex(selector)
+		rss = append(rss, rs)
```
Comment (jakubdyszkiewicz) [inline pkg/xds/envoy/listeners/fault_injection_configurer.go:72]: `rss = append(rss, rs)` what rss and rs stands for? Can we use more descriptive names?

File: pkg/xds/envoy/routes/tags_configurer_test.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+			tags: map[string]map[string]bool{
+				"tag2": {"value21": true},
+				"tag3": {"value31": true, "value32": true, "value33": true},
+				"tag1": {"value11": true, "value12": true},
+			},
+			expected: `
+            requestHeadersToAdd:
+            - header:
+                key: x-kuma-tags
+                value: '&tag1=value11,value12&&tag2=value21&&tag3=value31,value32,value33&'`,
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/xds/envoy/routes/tags_configurer_test.go:46]: With regex we have to care about 2 important things:
- we don't match the suffix of the _Key_
- we don't match the prefix of the _Value_

The first condition is covered by `&` before each _Key_. The second one is covered by `[,&]` after each _Value_.
So in the case of a single ampersand, we might match it with the end of the _Value_ and we have nothing to match with the beginning of the next _Key_.

Also, that could be solved by an additional comma after the last value - `&tag1=value1,value2,&tag2=value1&`. Probably additional comma will be more understandable, what do you think?

File: pkg/xds/envoy/tags/serialize.go (hunk 5 above)
Comment (lobkovilya) [reply pkg/xds/envoy/tags/serialize.go:15]: Sure, there were `return "&" + strings.Join(pairs, "&") + "&"` and I forgot to delete that `if`

File: pkg/xds/envoy/listeners/fault_injection_configurer.go (hunk 7 above)
Comment (lobkovilya) [reply pkg/xds/envoy/listeners/fault_injection_configurer.go:41]: I discovered that `HeaderMatcher` (at least with the same header's name) works as AND. We also have an option to generate separate fault filters, but that's probably bulkier solution

Comment (jakubdyszkiewicz) [reply pkg/xds/envoy/listeners/fault_injection_configurer.go:41]: Thanks, you are right
"A match will happen if all the headers in the config are present in the request with the same values (or based on presence if the value field is not in the config)."
https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/http/fault/v2/fault.proto

I'd lean towards picking a solution with generating many fault filters just to avoid crazy long regexes (btw. is 500 the limit of chars in regex?) and also for easier migration to `x-kuma-tag-*` in the future, but we can stick to or regex solution as well for now.

Comment (lobkovilya) [reply pkg/xds/envoy/listeners/fault_injection_configurer.go:41]: This is not exactly chars in regex:

_This field controls the RE2 “program size” which is a rough estimate of how complex a compiled regex is to evaluate. A regex that has a program size greater than the configured value will fail to compile. In this case, the configured max program size can be increased or the regex can be simplified. If not specified, the default is 100._

But there is a linear dependency between the size in chars and "program size". 
In both cases (many filters or complex regex) we will be limited by that "program size" and limited with number of tags/selectors in the policy. 

I'd stay with long regex for this PR and later we can compare which approach is faster - e.g. 10 filter faults with program size 500 or one filter fault with program size 5000

Comment (lobkovilya) [reply pkg/xds/envoy/listeners/fault_injection_configurer.go:41]: Oh and also we probably can calculate "program size" on the fly

File: pkg/core/policy/dataplane_matcher.go [hunk 10]
```diff
@@ -3,17 +3,18 @@ package policy
...
+	mesh_core "github.com/Kong/kuma/pkg/core/resources/apis/mesh"
+
 	mesh_proto "github.com/Kong/kuma/api/mesh/v1alpha1"
-	"github.com/Kong/kuma/pkg/core/resources/apis/mesh"
 )
 
-// SelectDataplanePolicy given a Dataplane definition and a list of DataplanePolicy returns the "best matching" DataplanePolicy.
-// A DataplanePolicy is considered a match if one of the inbound interfaces of a Dataplane or tag section on Gateway Dataplane has all tags of DataplanePolicy's selector.
+// SelectDataplanePolicy given a something that could be matched (mesh_proto.Dataplane,
+// mesh_proto.Dataplane_Networking_Inbound, etc.) and a list of DataplanePolicy returns the "best matching" DataplanePolicy.
```
Comment (jakubdyszkiewicz) [inline pkg/core/policy/dataplane_matcher.go:12]: Comment should be reverted

File: pkg/xds/envoy/tags/match_test.go (hunk 6 above)
Comment (jakubdyszkiewicz) [reply pkg/xds/envoy/tags/match_test.go:52]: I can't see the change

File: pkg/xds/server/components.go [hunk 11]
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"time"
 
+	"github.com/Kong/kuma/pkg/core/faultinjections"
```
Comment (jakubdyszkiewicz) [inline pkg/xds/server/components.go:7]: I think this should be in a group with other kuma imports. This goimport linter is not ideal unfortunately :(

File: pkg/xds/envoy/tags/match_test.go (hunk 6 above)
Comment (lobkovilya) [reply pkg/xds/envoy/tags/match_test.go:52]: I added a new test case lower:
```
Entry("match the latter valuer", testCase{
	serviceTags: mesh_proto.MultiValueTagSet{
		"tag1": {"value1": true, "value2": true},
	},
	selector: mesh_proto.SingleValueTagSet{
		"tag1": "value2",
	},
	expected: true,
}),
```
